### DISCUSSION TITLE
How to embed source files

### DISCUSSION THREAD

**Discussion by ocupe** [markdoc/markdoc#447] [Q&A] [answered]
Hello and many thanks for this nice piece of software!

I'm trying to embed source files into my documentation and was wondering what the best way to do this is.
Should I use partials? Something like this just returns an empty fence code block: (`/` before the fence ``` is just to escape them :) )

```mdoc
<!-- in doc-page.mdoc -->

# This is a docs page
The following code is embedded from example-code.ts
/```
{% partial file="example-code.ts" /%}
/```
```

Also setting `{% process=false %}` false on the fence does not work for me.
Writing a custom tag wouldn't be a problem, but I'm curious if I'm missing something obvious or maybe someone has a better idea or a working implementation they want to share.

**Comment by rpaul-stripe** [comment, marked as answer]
The `partial` tag basically treats the `file` attribute as a key and looks for a corresponding property in your Markdoc config (`config.partials[file]`) which it expects to be a parsed Markdoc AST. It's not really designed for raw, plain text content inclusion. I think you'd likely be better off making a custom tag for this.

**Reply by ocupe** [reply]
Thanks for the response 🙏. 
I ended up doing something that feels a little naughty but works for me.
I sandwich the file-content-string between ``` and then add it as a partial.

```ts
const rawSource = await fs.promises.readFile(filename, 'utf-8');
      
const source = '```ts\n' + rawSource + '\n```';

const tokens = tokenizer.tokenize(source);
const ast = Markdoc.parse(tokens);
```

Now I can use it like a normal partial and it will be rendered with my `Fence` component. :)

**Reply by segphault** [reply]
@Ocupe there may be things in your `rawSource` content that the Markdoc parser will try to operate on unexpectedly. What I would suggest instead is to programmatically create an AST with a `fenced` node that includes the `rawSource` as its content:

```js
const ast = new Markdoc.Ast.Node('document', {}, [
  new Markdoc.Ast.Node('fence', {
    language: 'ts',
    content: rawSource,
    process: false
  })
])
```

This lets you still treat the typescript file as a partial and it works basically the same way, but it avoids needing to use the parser, so you won't have to worry about sanitization.

**Reply by ocupe** [reply]
@segphault oh, that is much nicer! I tried `new Node('fence', ...` but that did not work. 
Perfect, this makes me happy, I'll use this! 👍
